### ISSUE TITLE
UDP client observing packet loss

### ISSUE THREAD

**Issue by ankitkothana** [netty/netty#9793]
### Expected behavior
UDP client should receive all the responses for all the requests.

### Actual behavior
Client is unable to get responses for all the requests sent.

### Minimal yet complete reproducer code (or URL to code)
`
       EventLoopGroup workerGroup = new NioEventLoopGroup();

        bootstrap = new Bootstrap();
        bootstrap.group(workerGroup);
        bootstrap.channel(NioDatagramChannel.class);
        bootstrap.option(ChannelOption.SO_RCVBUF, UDP_BUF_SIZE);
        bootstrap.option(ChannelOption.SO_SNDBUF, UDP_BUF_SIZE);
        bootstrap.option(ChannelOption.SO_BROADCAST, true);
        bootstrap.option(ChannelOption.SO_REUSEADDR, true);

        channel = bootstrap.bind(new InetSocketAddress(0)).channel();

        public void sendUDP(Message message, InetSocketAddress host) throws InterruptedException {
        channel.writeAndFlush(new DefaultAddressedEnvelope<Message, SocketAddress>(message, host));
    }
`

### Netty version
4.1.13

### JVM version (e.g. `java -version`)
java 8

### OS version (e.g. `uname -a`)
centos 7

Could you please suggest what could be possible reason that ChannelInboundHandler is unable to read all the responses?

**Comment by ankitkothana** [comment]
Tried with other clients from same host to the same host and no response loss is observed. However with the above client we are getting almost 50% response loss with **bulk requests**.

**Comment by normanmaurer** [comment]
@ankitkothana first of you should upgrade your netty version as it is super old. Second you may need to use a different `FixedSizeRecvByteBufAllocator` and set the buffer size bigger.

**Comment by ankitkothana** [comment]
As per https://mvnrepository.com/artifact/io.netty/netty-all 4.1.43.Final is the latest version and that is what we are using. The buffer size is already set to 16 MB. Do we need to increase it further?

**Comment by normanmaurer** [comment]
@ankitkothana now I am confused... in the issue you said `4.1.13` ?

**Comment by normanmaurer** [comment]
@ankitkothana I think you want to do:

```
bootstrap.option(ChannelOption.RCVBUF_ALLOCATOR, new FixedRecvByteBufAllocator(UDP_BUF_SIZE)); 
```

That said 16MB over UDP will never work so I am not sure why you set this. Even with Jumbo packets I have a hard time believing you will ever see bigger packets then 9kb.

**Comment by ankitkothana** [comment]
Sorry for adding a wrong version number in the issue. I have updated the issue.

`private static final int UDP_BUF_SIZE = 9216;`

Reduced buffer size to 9KB and updated the bootstrap options:

`
        bootstrap.option(ChannelOption.SO_RCVBUF, UDP_BUF_SIZE);
        bootstrap.option(ChannelOption.SO_SNDBUF, UDP_BUF_SIZE);
        bootstrap.option(ChannelOption.SO_BROADCAST, true);
        bootstrap.option(ChannelOption.SO_REUSEADDR, true);
        bootstrap.option(ChannelOption.RCVBUF_ALLOCATOR, new FixedRecvByteBufAllocator(UDP_BUF_SIZE)); `

Still I see a huge packet loss.

**Comment by normanmaurer** [comment]
Without a reproducer it is hard to help ... sorry

**Comment by ankitkothana** [comment]
Here is the complete code:

https://github.com/ankitkothana/client.git

Please have a look into it.

**Comment by ankitkothana** [comment]
Any updates on this?

**Comment by kapilmaheshwari** [comment]
Facing the same issue.

**Comment by normanmaurer** [comment]
@ankitkothana I wonder if you are just writing too fast and so the remote peer will just drop stuff. Can you add some `Thread.sleep(50);` in your for / while loop ? It is not uncommon for DNS servers to rate limit clients.

**Comment by ankitkothana** [comment]
Usually DNS servers have a QPS of min 50K over UDP. I have tried hitting the DNS server with dnsperf (C++ utility for performance testing of DNS servers) and observed no packet loss even at a higher request rate per second. This is not a server side issue.

Event tried to add a sleep of 50 ms after 100 requests. Packet loss was still > 40% of actual requests.

The only workaround that worked for us is to add a sleep of 1 ms after every request. This gave us a success rate of 99%. 

Could you please explain why response packets are dropped without sleep interval? Is this a bug in netty?

**Comment by morvael** [comment]
Maybe my observations will help you:
1) UDP is giving no guarantee about all packets reaching target
2) Writing without any limits leads to filling network socket buffers (this may not be an accurate term). I have an app that pushes messages to clients as fast as it can (it's on TCP though), and it has to detect that buffers are full, and stop writing until buffers are free again. You have to write each message checking channel.isWritable() before, and if it's not you have to stop until you get notified in channelWritabilityChanged method that it's writable again:
```
    @Override
    public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
        if (ctx.channel().isWritable()) {
            sendMessagesInLoop(ctx);
        }
    }

    private void sendMessagesInLoop(ChannelHandlerContext ctx) {
        while (ctx.channel().isWritable()) {
          //send
        }
    }
```
You can either call sendMessagesInLoop for the first time after connecting or (my preferred way) by using IdleStateHandler in the pipeline, and listening to IdleStateEvent with WRITER_IDLE state in userEventTriggered.

**Comment by ankitkothana** [comment]
Thanks @morvael  for providing useful observations. Can you provide an example of how to use IdleStateHandler and handle the WRITER_IDLE event.

**Comment by morvael** [comment]
Yes, it should be used for heartbeats, but it also works for me as a way to know when to start writing again. Extend io.netty.channel.ChannelInitializer<Channel> and write something like that:
```
    @Override
    protected void initChannel(Channel ch) throws Exception {
        ChannelPipeline pipe = ch.pipeline();
        if ((readIdleMillis > 0) || (writeIdleMillis > 0)) {
            pipe.addLast("idleHandler", new io.netty.handler.timeout.IdleStateHandler(readIdleMillis, writeIdleMillis, 0, TimeUnit.MILLISECONDS));
```
...then add your other handlers after (or experiment with other add methods to put it higher upstream than your code checking channelWritabilityChanged). Then give your ChannelInitializer to bootstrap as handler.

**Comment by morvael** [comment]
Sorry, it seems I have confused two things. IdleStateHandler is indeed better used to heartbeats. You don't need it. Just overrride channelWritabilityChanged in your ChannelInboundHandlerAdapter, as in my first example.

**Comment by morvael** [comment]
In your case you can start writing as the client connects, pausing when channel is no longer writable, and starting again when it's writable again. As for using IdleStateHandler, now that it's in your pipeline you can use:
```
    @Override
    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
        if (evt instanceof IdleStateEvent) {
            IdleStateEvent e = (IdleStateEvent) evt;
            if (e.state() == IdleState.WRITER_IDLE) {
                sendMessagesInLoop(ctx);
            }
        }
    }
```
in your ChannelInboundHandlerAdapter.

**Comment by ankitkothana** [comment]
Thanks @morvael . Incorporating these suggestions didn't help though.

DNSClientIntializerTemp.java

``` java
@Override
    protected void initChannel(DatagramChannel ch) throws Exception {
        ChannelPipeline channelPipeline = ch.pipeline();
        channelPipeline.addLast("MessageDecoder",
                new DatagramPacketDecoder(new DNSDecoder()));
        channelPipeline.addLast("MessageEncoder",
                new DatagramPacketEncoder<>(new DNSEncoder()));
        channelPipeline.addLast("idleHandler", new IdleStateHandler(0, 1, 0));
        channelPipeline.addLast("MesageHandler", new MessageHandler());
    }
```
MessageHandler.java

``` java
@Override
    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
        if (evt instanceof IdleStateEvent) {
            IdleStateEvent e = (IdleStateEvent) evt;
            if (e.state() == IdleState.WRITER_IDLE) {
                sendMessagesInLoop(ctx);
            }
        }
    }
    
    private void sendMessagesInLoop(ChannelHandlerContext ctx) {
        while (ctx.channel().isWritable()) {
          
        }
    }
```
Is there something else I am missing. I still see a huge packet loss on increasing concurrency and number of requests. 

Even tried to add logger in channelWritabilityChanged method. It seems like writability is never changed and we are able to push the messages without buffer overflow.

**Comment by ankitkothana** [comment]
@normanmaurer @morvael any updates would be helpful here. We are seeing lots of issues due to packet loss.

**Comment by hyperxpro** [comment]
I am sure that it's not Netty problem. My Company has created its own DNS Resolver and Authoritative Server on top of Netty.

Resolver can easily process 500K RPS and Authoritative hits upto 1.2M RPS with 0% Packet Loss (10G LAN). We did face packet loss issue on Linux but easily fixed it with few Kernel tuning hacks.

I'd suggest testing your Client and Server using iperf3 and check for possible packet loss.

**Comment by TreviD** [comment]
I have the same problem! But I solved this problem by adding `sync()` after `writeAndFlush` method.
So can someone explain why?
```
public void send(String domainName) throws InterruptedException {
        DatagramDnsQuery dnsQuery = new DatagramDnsQuery(null, new InetSocketAddress("127.0.0.1", 53), threadLocalRandom.nextInt(0xFFFF));
        dnsQuery.addRecord(DnsSection.QUESTION, new DefaultDnsQuestion(domainName, DnsRecordType.A));
        channel.writeAndFlush(dnsQuery).sync();
}

```

**Comment by chrisvest** [comment]
@TreviD `sync()` will wait for the operation to complete, which here means the DNS query was sent.
